server/sonar-db-dao/src/main/java/org/sonar/db/property/InternalPropertiesDao.java

```diff
@@ -77,13 +90,57 @@ public void saveAsEmpty(DbSession dbSession, String key) {
   }
 
   /**
+   * @return a Map with an {link Optional<String>} for each String in {@code keys}.
+   */
+  public Map<String, Optional<String>> selectByKeys(DbSession dbSession, Set<String> keys) {
+    if (keys == null || keys.isEmpty()) {
+      return Collections.emptyMap();
+    }
+    if (keys.size() == 1) {
+      String key = keys.iterator().next();
+      return ImmutableMap.of(key, selectByKey(dbSession, key));
+    }
+    keys.forEach(InternalPropertiesDao::checkKey);
+
+    InternalPropertiesMapper mapper = getMapper(dbSession);
+    List<InternalPropertyDto> res = mapper.selectAsText(ImmutableList.copyOf(keys));
+    Map<String, Optional<String>> builder = new HashMap<>(keys.size());
+    res.forEach(internalPropertyDto -> {
+      String key = internalPropertyDto.getKey();
+      if (internalPropertyDto.isEmpty()) {
+        builder.put(key, OPTIONAL_OF_EMPTY_STRING);
+      }
+      if (internalPropertyDto.getValue() != null) {
+        builder.put(key, Optional.of(internalPropertyDto.getValue()));
+      }
+    });
+    // return Optional.empty() for all keys without a DB entry
+    Sets.difference(keys, res.stream().map(InternalPropertyDto::getKey).collect(Collectors.toSet()))
+      .forEach(key -> builder.put(key, Optional.empty()));
+    // keys for which there isn't a text or empty value found yet
+    List<String> keyWithClobValue = ImmutableList.copyOf(Sets.difference(keys, builder.keySet()));
```

Comment:
This section of the code is hard to read. It seems that we already already put all keys passed as argument into the builder at this point.
Perhaps having a single for loop through all keys given as argument would make it easier to read.

Reply:
incorrect. We only put in builder keys which:
1. had a row in table
2. had a flag empty=true or a value stored as text

then we put an empty() for each argument which has no row in DB

then we put a value for each argument with a clob value

finally, in case of inconsistency in DB, we put an empty for argument which has a row in DB, but neither empty=true, nor a text value, nor a clob value